ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/StateTreeManager.java

```diff
@@ -67,117 +58,71 @@ public void onChainStarted(ChainStartEvent event) {
-        LOG.info("block slot: " + block.get().getSlot());
-        LOG.info("block state root: " + blockStateRoot.toHexString());
-        newState = BeaconState.deepCopy(parentState);
-
-        // run stateTransition.initiate() on empty slots from parentBlock.slot to block.slot-1
-        int counter = 0;
-        while (newState.getSlot().compareTo(UnsignedLong.valueOf(block.get().getSlot() - 1)) < 0) {
-          if (counter == 0) {
-            LOG.info(
-                "Transitioning state from slot: "
-                    + newState.getSlot()
-                    + " to slot: "
-                    + UnsignedLong.valueOf(block.get().getSlot() - 1));
-          }
-          stateTransition.initiate(newState, null, store);
-          newStateRoot = HashTreeUtil.hash_tree_root(newState.toBytes());
-          this.store.addState(newStateRoot, newState);
-          newState = BeaconState.deepCopy(newState);
-          counter++;
-        }
+    List<Optional<BeaconBlock>> unprocessedBlocks =
+        this.store.getUnprocessedBlocksUntilSlot(nodeSlot);
+    unprocessedBlocks.forEach((block) -> processFork(block));
 
-        // run stateTransition.initiate() on block.slot
-        stateTransition.initiate(newState, block.get(), store);
-        newStateRoot = HashTreeUtil.hash_tree_root(newState.toBytes());
-
-        // state root verification
-        if (blockStateRoot.equals(newStateRoot)) {
-          LOG.info("The block's state root matches the calculated state root!");
-          LOG.info("  new state root: " + newStateRoot.toHexString());
-          LOG.info("  block state root: " + blockStateRoot.toHexString());
-          // TODO: storing block and state together as a tuple would be more convenient
-          this.store.addProcessedBlock(blockStateRoot, block.get());
-          this.store.addProcessedBlock(blockRoot, block.get());
-          this.store.addState(newStateRoot, newState);
-
-          // run stateTransition.initiate() on slots from block.slot to node.slot
-          counter = 0;
-          while (newState.getSlot().compareTo(nodeSlot) < 0) {
-            if (counter == 0) {
-              LOG.info(
-                  "Transitioning state from slot: " + newState.getSlot() + " to slot: " + nodeSlot);
-            }
-            newState = BeaconState.deepCopy(newState);
-            stateTransition.initiate(newState, null, store);
-            newStateRoot = HashTreeUtil.hash_tree_root(newState.toBytes());
-            this.store.addState(newStateRoot, newState);
-            counter++;
-          }
-          LOG.info("latest state root: " + newStateRoot.toHexString());
-          this.canonical_state = newState;
-        } else {
-          LOG.info("The block's state root does not match the calculated state root!");
-          LOG.info("  new state root: " + newStateRoot.toHexString());
-          LOG.info("  block state root: " + blockStateRoot.toHexString());
-          shouldProcessBlock = false;
```

Comment:
if the state roots don't match don't we need to resort to a state transition with no block?  this could happen if the block proposer generates an invalid block